{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c68c74af",
   "metadata": {},
   "source": [
    "%%capture install_messages\n",
    "!pip install pydna\n",
    "!pip install Bio\n",
    "!pip install teemi\n",
    "!pip install pyqt5\n",
    "!pip install dnaplotlib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d29017",
   "metadata": {},
   "source": [
    "## GFP Assay combinatorics\n",
    "### Load promoter & Terminator library and GFP sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d49c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from teemi.design.combinatorial_design import DesignAssembly\n",
    "from src.smart_functions import read_fasta_to_dseqrecords\n",
    "from IPython.display import display\n",
    "\n",
    "import os\n",
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f612438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Promoters: 9, CDS: 1, Terminators: 9\n",
      "Example promoter length: 1000\n",
      "Example CDS length: 714\n"
     ]
    }
   ],
   "source": [
    "promoters_fa = r'data/promoter_terminator_library/promoters.fasta'\n",
    "terminators_fa = r'data/promoter_terminator_library/terminators.fasta'\n",
    "\n",
    "m_paba_fa = r'data/insert_sequences/PABA.fasta'\n",
    "dTomato_fa = r'data/insert_sequences/dTomato_non_optimized.fasta'\n",
    "\n",
    "promoters, promoter_names = read_fasta_to_dseqrecords(promoters_fa)\n",
    "cds_records, cds_names     = read_fasta_to_dseqrecords(dTomato_fa)\n",
    "terminators, terminator_names = read_fasta_to_dseqrecords(terminators_fa)\n",
    "m_paba, m_paba_names = read_fasta_to_dseqrecords(m_paba_fa)\n",
    "\n",
    "\n",
    "list_of_seqs  = [[m_paba[0]], promoters, [cds_records[0]], terminators, [m_paba[1]]]\n",
    "\n",
    "print(f\"Promoters: {len(promoters)}, CDS: {len(cds_records)}, Terminators: {len(terminators)}\")\n",
    "if promoters:\n",
    "    print(\"Example promoter length:\", len(promoters[0].seq))\n",
    "if cds_records:\n",
    "    print(\"Example CDS length:\", len(cds_records[0].seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af486ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variants: 81 saved -> c:\\Users\\alexa\\OneDrive - Danmarks Tekniske Universitet\\27460 - Syntesebiologi\\Exam\\Combinatorics\\variants_library.csv\n",
      "Primers:  51 saved -> c:\\Users\\alexa\\OneDrive - Danmarks Tekniske Universitet\\27460 - Syntesebiologi\\Exam\\Combinatorics\\primers_list.csv\n",
      "PCRs:     90 saved -> c:\\Users\\alexa\\OneDrive - Danmarks Tekniske Universitet\\27460 - Syntesebiologi\\Exam\\Combinatorics\\pcr_plan.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>Systematic_name</th>\n",
       "      <th>Variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PKG1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>PKG1</td>\n",
       "      <td>(1, 1, 1)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PKG1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ADH1</td>\n",
       "      <td>(1, 1, 2)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PKG1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TDH3</td>\n",
       "      <td>(1, 1, 3)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PKG1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ACT</td>\n",
       "      <td>(1, 1, 4)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PKG1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TEF1</td>\n",
       "      <td>(1, 1, 5)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0        1     2 Systematic_name  Variant\n",
       "0  PKG1  reverse  PKG1       (1, 1, 1)        0\n",
       "1  PKG1  reverse  ADH1       (1, 1, 2)        1\n",
       "2  PKG1  reverse  TDH3       (1, 1, 3)        2\n",
       "3  PKG1  reverse   ACT       (1, 1, 4)        3\n",
       "4  PKG1  reverse  TEF1       (1, 1, 5)        4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>anneals to</th>\n",
       "      <th>sequence</th>\n",
       "      <th>annealing temperature</th>\n",
       "      <th>length</th>\n",
       "      <th>price(DKK)</th>\n",
       "      <th>description</th>\n",
       "      <th>footprint</th>\n",
       "      <th>len_footprint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P001</td>\n",
       "      <td>PKG1</td>\n",
       "      <td>(G, T, G, T, T, A, T, C, T, G, G, T, C, G, T, ...</td>\n",
       "      <td>54.88</td>\n",
       "      <td>19</td>\n",
       "      <td>34.2</td>\n",
       "      <td>Anneals to PKG1</td>\n",
       "      <td>(G, T, G, T, T, A, T, C, T, G, G, T, C, G, T, ...</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P002</td>\n",
       "      <td>PKG1</td>\n",
       "      <td>(C, T, C, C, T, C, G, C, C, C, T, T, G, G, A, ...</td>\n",
       "      <td>54.79</td>\n",
       "      <td>37</td>\n",
       "      <td>66.6</td>\n",
       "      <td>Anneals to PKG1, overlaps to 750bp_PCR_prod</td>\n",
       "      <td>(T, G, T, G, G, A, T, T, G, T, G, A, A, G, A, ...</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P003</td>\n",
       "      <td>reverse</td>\n",
       "      <td>(T, C, T, T, C, T, T, C, A, C, A, A, T, C, C, ...</td>\n",
       "      <td>54.10</td>\n",
       "      <td>32</td>\n",
       "      <td>57.6</td>\n",
       "      <td>Anneals to reverse, overlaps to PKG1</td>\n",
       "      <td>(A, T, G, T, C, C, A, A, G, G, G, C, G, A)</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P004</td>\n",
       "      <td>reverse</td>\n",
       "      <td>(A, C, T, T, G, T, T, C, A, T, T, T, T, G, C, ...</td>\n",
       "      <td>53.64</td>\n",
       "      <td>34</td>\n",
       "      <td>61.2</td>\n",
       "      <td>Anneals to reverse, overlaps to 1018bp_PCR_prod</td>\n",
       "      <td>(C, T, T, G, T, A, G, A, G, C, T, C, G, T, C, C)</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P005</td>\n",
       "      <td>PKG1</td>\n",
       "      <td>(A, T, G, G, A, C, G, A, G, C, T, C, T, A, C, ...</td>\n",
       "      <td>54.89</td>\n",
       "      <td>43</td>\n",
       "      <td>77.4</td>\n",
       "      <td>Anneals to PKG1, overlaps to reverse</td>\n",
       "      <td>(A, A, A, G, C, A, A, A, A, T, G, A, A, C, A, ...</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     id anneals to                                           sequence  \\\n",
       "0  P001       PKG1  (G, T, G, T, T, A, T, C, T, G, G, T, C, G, T, ...   \n",
       "1  P002       PKG1  (C, T, C, C, T, C, G, C, C, C, T, T, G, G, A, ...   \n",
       "2  P003    reverse  (T, C, T, T, C, T, T, C, A, C, A, A, T, C, C, ...   \n",
       "3  P004    reverse  (A, C, T, T, G, T, T, C, A, T, T, T, T, G, C, ...   \n",
       "4  P005       PKG1  (A, T, G, G, A, C, G, A, G, C, T, C, T, A, C, ...   \n",
       "\n",
       "   annealing temperature  length  price(DKK)  \\\n",
       "0                  54.88      19        34.2   \n",
       "1                  54.79      37        66.6   \n",
       "2                  54.10      32        57.6   \n",
       "3                  53.64      34        61.2   \n",
       "4                  54.89      43        77.4   \n",
       "\n",
       "                                       description  \\\n",
       "0                                  Anneals to PKG1   \n",
       "1      Anneals to PKG1, overlaps to 750bp_PCR_prod   \n",
       "2             Anneals to reverse, overlaps to PKG1   \n",
       "3  Anneals to reverse, overlaps to 1018bp_PCR_prod   \n",
       "4             Anneals to PKG1, overlaps to reverse   \n",
       "\n",
       "                                           footprint  len_footprint  \n",
       "0  (G, T, G, T, T, A, T, C, T, G, G, T, C, G, T, ...             19  \n",
       "1  (T, G, T, G, G, A, T, T, G, T, G, A, A, G, A, ...             19  \n",
       "2         (A, T, G, T, C, C, A, A, G, G, G, C, G, A)             14  \n",
       "3   (C, T, T, G, T, A, G, A, G, C, T, C, G, T, C, C)             16  \n",
       "4  (A, A, A, G, C, A, A, A, A, T, G, A, A, C, A, ...             25  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pcr_number</th>\n",
       "      <th>template</th>\n",
       "      <th>forward_primer</th>\n",
       "      <th>reverse_primer</th>\n",
       "      <th>f_tm</th>\n",
       "      <th>r_tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PCR1</td>\n",
       "      <td>PKG1</td>\n",
       "      <td>P001</td>\n",
       "      <td>P002</td>\n",
       "      <td>54.88</td>\n",
       "      <td>54.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PCR2</td>\n",
       "      <td>reverse</td>\n",
       "      <td>P003</td>\n",
       "      <td>P004</td>\n",
       "      <td>54.10</td>\n",
       "      <td>53.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PCR3</td>\n",
       "      <td>PKG1</td>\n",
       "      <td>P005</td>\n",
       "      <td>P006</td>\n",
       "      <td>54.89</td>\n",
       "      <td>55.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PCR4</td>\n",
       "      <td>reverse</td>\n",
       "      <td>P003</td>\n",
       "      <td>P007</td>\n",
       "      <td>54.10</td>\n",
       "      <td>53.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PCR5</td>\n",
       "      <td>ADH1</td>\n",
       "      <td>P008</td>\n",
       "      <td>P009</td>\n",
       "      <td>54.24</td>\n",
       "      <td>53.28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pcr_number template forward_primer reverse_primer   f_tm   r_tm\n",
       "0       PCR1     PKG1           P001           P002  54.88  54.79\n",
       "1       PCR2  reverse           P003           P004  54.10  53.64\n",
       "2       PCR3     PKG1           P005           P006  54.89  55.38\n",
       "3       PCR4  reverse           P003           P007  54.10  53.64\n",
       "4       PCR5     ADH1           P008           P009  54.24  53.28"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "TARGET_TM = 65\n",
    "LIMIT = 13\n",
    "OVERLAP = 35\n",
    "\n",
    "design = DesignAssembly(list_of_seqs, list_of_pads=[], positions_of_pads=[], target_tm=TARGET_TM, limit=LIMIT, overlap=OVERLAP)\n",
    "variants_df = design.show_variants_lib_df()          \n",
    "primers_df  = design.primer_list_to_dataframe()      \n",
    "pcrs_df = design.pcr_list_to_dataframe()\n",
    "\n",
    "\n",
    "out_dir = os.getcwd()\n",
    "variants_csv = os.path.join(out_dir, \"flouresence_variants_library.csv\")\n",
    "primers_csv  = os.path.join(out_dir, \"flouresence_primers_list.csv\")\n",
    "pcrs_csv= os.path.join(out_dir, \"flouresence_pcr_plan.csv\")\n",
    "\n",
    "variants_df.to_csv(variants_csv, index=False)\n",
    "primers_df.to_csv(primers_csv, index=False)\n",
    "pcrs_df.to_csv(pcrs_csv, index=False)\n",
    "\n",
    "print(f\"Variants: {len(variants_df)} saved -> {variants_csv}\")\n",
    "print(f\"Primers:  {len(primers_df)} saved -> {primers_csv}\")\n",
    "print(f\"PCRs:     {len(pcrs_df)} saved -> {pcrs_csv}\")\n",
    "display(variants_df.head())\n",
    "display(primers_df.head())\n",
    "display(pcrs_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "717c7289",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2644.2\n"
     ]
    }
   ],
   "source": [
    "print(primers_df[\"price(DKK)\"].sum())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
