{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bbe04a4e",
   "metadata": {},
   "source": [
    "# Creation of in vitro insertion parts for pfa chain in C.Cinerea"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d453a96d",
   "metadata": {},
   "source": [
    "Now that promoters have been collected and generated and ran through a quantative flouresence analysis we can create the full list of inserts and necesarry primers for the top 5 primers and terminators. \n",
    "\n",
    "This insert will contain: <br>\n",
    "- PABA marker gene <br>\n",
    "- PUFA synthesis pathway <br>\n",
    "    - pfa1,2,3 and a pptase <br>\n",
    "- Hygromicin marker gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31a4f34c",
   "metadata": {},
   "source": [
    "*this is where we insert a cool image of the creation*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd194758",
   "metadata": {},
   "source": [
    "## Loading Libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff6a4f2",
   "metadata": {},
   "source": [
    "## Genetic elements"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8af7b9c4",
   "metadata": {},
   "source": [
    "### HR Domain"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd1fa971",
   "metadata": {},
   "source": [
    "The insertion site into C.Cinerea will be the spoII site based on the report: <br> *Non-conventional fungi are efficient heterologous hosts for natural product production.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38697c03",
   "metadata": {},
   "source": [
    "### PABA marker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27ad213",
   "metadata": {},
   "source": [
    "### PUFA Synthesis pathway"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aa0bead",
   "metadata": {},
   "source": [
    "### Hygromicin marker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e3b7bc",
   "metadata": {},
   "source": [
    "## Primer creation using TEEMI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4164e4d3",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bb407a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from pydna.dseqrecord import Dseqrecord\n",
    "from pydna.assembly import Assembly\n",
    "from teemi.design.combinatorial_design import DesignAssembly\n",
    "\n",
    "import pandas as pd\n",
    "import dnaplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from IPython.display import display\n",
    "import IPython.core.display\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e4e07f",
   "metadata": {},
   "source": [
    "### Fetching everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb531935",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/data/insert_sequences/promoters'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 10\u001b[0m\n\u001b[0;32m      7\u001b[0m m_hygro_fa \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnotebooks\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mdata\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124minsert_sequnces\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mHygromycin.fasta\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;66;03m#hygromicin including promoter and terminator.\u001b[39;00m\n\u001b[0;32m      8\u001b[0m cds_fa \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnotebooks\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mdata\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124minsert_sequnces\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124mpufa_optimized.fasta\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;66;03m#pfa123 and pptase\u001b[39;00m\n\u001b[1;32m---> 10\u001b[0m promoters, promoter_names \u001b[38;5;241m=\u001b[39m \u001b[43mread_fasta_to_dseqrecords\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtop_promoters\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     11\u001b[0m terminators, terminator_names \u001b[38;5;241m=\u001b[39m read_fasta_to_dseqrecords(top_terminators)\n\u001b[0;32m     13\u001b[0m cds_records, cds_names \u001b[38;5;241m=\u001b[39m read_fasta_to_dseqrecords(cds_fa)\n",
      "File \u001b[1;32mc:\\Users\\Bruger\\Desktop\\GIT-Projects\\Synthetic biology\\27460_synthetic_promoters\\notebooks\\smart_functions.py:8\u001b[0m, in \u001b[0;36mread_fasta_to_dseqrecords\u001b[1;34m(fasta_path)\u001b[0m\n\u001b[0;32m      6\u001b[0m recs \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m      7\u001b[0m names \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m----> 8\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m \u001b[43mSeqIO\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfasta_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m:\n\u001b[0;32m      9\u001b[0m     seq \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(r\u001b[38;5;241m.\u001b[39mseq)\u001b[38;5;241m.\u001b[39mupper()\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     10\u001b[0m     dr \u001b[38;5;241m=\u001b[39m Dseqrecord(seq)\n",
      "File \u001b[1;32mc:\\Users\\Bruger\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\__init__.py:626\u001b[0m, in \u001b[0;36mparse\u001b[1;34m(handle, format, alphabet)\u001b[0m\n\u001b[0;32m    624\u001b[0m iterator_generator \u001b[38;5;241m=\u001b[39m _FormatToIterator\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mformat\u001b[39m)\n\u001b[0;32m    625\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m iterator_generator:\n\u001b[1;32m--> 626\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43miterator_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    627\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mformat\u001b[39m \u001b[38;5;129;01min\u001b[39;00m AlignIO\u001b[38;5;241m.\u001b[39m_FormatToIterator:\n\u001b[0;32m    628\u001b[0m     \u001b[38;5;66;03m# Use Bio.AlignIO to read in the alignments\u001b[39;00m\n\u001b[0;32m    629\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m (r \u001b[38;5;28;01mfor\u001b[39;00m alignment \u001b[38;5;129;01min\u001b[39;00m AlignIO\u001b[38;5;241m.\u001b[39mparse(handle, \u001b[38;5;28mformat\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m alignment)\n",
      "File \u001b[1;32mc:\\Users\\Bruger\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\FastaIO.py:196\u001b[0m, in \u001b[0;36mFastaIterator.__init__\u001b[1;34m(self, source, alphabet)\u001b[0m\n\u001b[0;32m    194\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m alphabet \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    195\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe alphabet argument is no longer supported\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m--> 196\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfmt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mFasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m    197\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    198\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstream)\n",
      "File \u001b[1;32mc:\\Users\\Bruger\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\Interfaces.py:81\u001b[0m, in \u001b[0;36mSequenceIterator.__init__\u001b[1;34m(self, source, alphabet, fmt)\u001b[0m\n\u001b[0;32m     79\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(source, _PathLikeTypes):\n\u001b[0;32m     80\u001b[0m     mode \u001b[38;5;241m=\u001b[39m modes[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m---> 81\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     82\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m     83\u001b[0m     value \u001b[38;5;241m=\u001b[39m source\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;241m0\u001b[39m)\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/data/insert_sequences/promoters'"
     ]
    }
   ],
   "source": [
    "from smart_functions import read_fasta_to_dseqrecords\n",
    "\n",
    "top_promoters = r'/data/insert_sequences/promoters.fasta'\n",
    "top_terminators = r'/data/insert_sequences/promoters.fasta'\n",
    "\n",
    "m_paba_fa = r'notebooks\\data\\insert_sequnces\\PABA.fasta'\n",
    "m_hygro_fa = r'notebooks\\data\\insert_sequnces\\Hygromycin.fasta' #hygromicin including promoter and terminator.\n",
    "cds_fa = r'notebooks\\data\\insert_sequnces\\pufa_optimized.fasta' #pfa123 and pptase\n",
    "\n",
    "promoters, promoter_names = read_fasta_to_dseqrecords(top_promoters)\n",
    "terminators, terminator_names = read_fasta_to_dseqrecords(top_terminators)\n",
    "\n",
    "cds_records, cds_names = read_fasta_to_dseqrecords(cds_fa)\n",
    "m_paba, m_paba_names = read_fasta_to_dseqrecords(m_paba_fa)\n",
    "m_hygro, m_hygro_names = read_fasta_to_dseqrecords(m_hygro_fa)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f9ea94b",
   "metadata": {},
   "source": [
    "#### A quick count of sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87fdf520",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Promoters: {len(promoters)}, CDS: {len(cds_records)}, Terminators: {len(terminators)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8854ab9",
   "metadata": {},
   "source": [
    "### Putting it all together in a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8eeb4cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['hr1-PABA1-p1-cds1-t1-p1-cds2-t1-p1-cds3-t1-p1-cds4-t1-PABA2-hr2', 'hr1-PABA1-p1-cds1-t2-p1-cds2-t2-p1-cds3-t2-p1-cds4-t2-PABA2-hr2', 'hr1-PABA1-p1-cds1-t3-p1-cds2-t3-p1-cds3-t3-p1-cds4-t3-PABA2-hr2', 'hr1-PABA1-p1-cds1-t4-p1-cds2-t4-p1-cds3-t4-p1-cds4-t4-PABA2-hr2', 'hr1-PABA1-p2-cds1-t1-p2-cds2-t1-p2-cds3-t1-p2-cds4-t1-PABA2-hr2', 'hr1-PABA1-p2-cds1-t2-p2-cds2-t2-p2-cds3-t2-p2-cds4-t2-PABA2-hr2', 'hr1-PABA1-p2-cds1-t3-p2-cds2-t3-p2-cds3-t3-p2-cds4-t3-PABA2-hr2', 'hr1-PABA1-p2-cds1-t4-p2-cds2-t4-p2-cds3-t4-p2-cds4-t4-PABA2-hr2', 'hr1-PABA1-p3-cds1-t1-p3-cds2-t1-p3-cds3-t1-p3-cds4-t1-PABA2-hr2', 'hr1-PABA1-p3-cds1-t2-p3-cds2-t2-p3-cds3-t2-p3-cds4-t2-PABA2-hr2', 'hr1-PABA1-p3-cds1-t3-p3-cds2-t3-p3-cds3-t3-p3-cds4-t3-PABA2-hr2', 'hr1-PABA1-p3-cds1-t4-p3-cds2-t4-p3-cds3-t4-p3-cds4-t4-PABA2-hr2', 'hr1-PABA1-p4-cds1-t1-p4-cds2-t1-p4-cds3-t1-p4-cds4-t1-PABA2-hr2', 'hr1-PABA1-p4-cds1-t2-p4-cds2-t2-p4-cds3-t2-p4-cds4-t2-PABA2-hr2', 'hr1-PABA1-p4-cds1-t3-p4-cds2-t3-p4-cds3-t3-p4-cds4-t3-PABA2-hr2', 'hr1-PABA1-p4-cds1-t4-p4-cds2-t4-p4-cds3-t4-p4-cds4-t4-PABA2-hr2']\n",
      "16 constructs generated\n"
     ]
    }
   ],
   "source": [
    "full_constructs = []\n",
    "\n",
    "for p in promoters:\n",
    "    for t in terminators:\n",
    "        \n",
    "        # Build the entire block of all CDSs wrapped in p and t\n",
    "        cds_block = []\n",
    "        for cds in cds_records:\n",
    "            cds_block.extend([p, cds, t])\n",
    "\n",
    "        # Now wrap with HR and PABA components\n",
    "        construct = [\n",
    "            HR[0],\n",
    "            PABA[0],\n",
    "            *cds_block,\n",
    "            PABA[1],\n",
    "            HR[1]\n",
    "        ]\n",
    "        \n",
    "        # Join into final construct string\n",
    "        full_constructs.append(\"-\".join(construct))\n",
    "\n",
    "print(full_constructs)\n",
    "print(len(full_constructs), \"constructs generated\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ba745a9",
   "metadata": {},
   "source": [
    "### Combining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d9a4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TARGET_TM = 55\n",
    "# LIMIT = 13\n",
    "# OVERLAP = 35\n",
    "\n",
    "\n",
    "\n",
    "# design = DesignAssembly(list_of_seqs, list_of_pads=[], positions_of_pads=[], target_tm=TARGET_TM, limit=LIMIT, overlap=OVERLAP)\n",
    "#variants_df = design.show_variants_lib_df()          \n",
    "#primers_df  = design.primer_list_to_dataframe()      \n",
    "#pcrs_df = design.pcr_list_to_dataframe()\n",
    "\n",
    "\n",
    "#out_dir = os.path.abspath(\".\")\n",
    "#variants_csv = os.path.join(out_dir, \"variants_library.csv\")\n",
    "#primers_csv  = os.path.join(out_dir, \"primers_list.csv\")\n",
    "#pcrs_csv= os.path.join(out_dir, \"pcr_plan.csv\")\n",
    "\n",
    "#variants_df.to_csv(variants_csv, index=False)\n",
    "#primers_df.to_csv(primers_csv, index=False)\n",
    "#pcrs_df.to_csv(pcrs_csv, index=False)\n",
    "\n",
    "#print(f\"Variants: {len(variants_df)} saved -> {variants_csv}\")\n",
    "#print(f\"Primers:  {len(primers_df)} saved -> {primers_csv}\")\n",
    "#print(f\"PCRs:     {len(pcrs_df)} saved -> {pcrs_csv}\")\n",
    "#display(variants_df.head())\n",
    "#display(primers_df.head())\n",
    "#display(pcrs_df.head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
