{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bbe04a4e",
   "metadata": {},
   "source": [
    "# Creation of in vitro insertion parts for pfa chain in C.Cinerea"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d453a96d",
   "metadata": {},
   "source": [
    "Now that promoters have been collected and generated and ran through a quantative flouresence analysis we can create the full list of inserts and necesarry primers for the top 5 primers and terminators. \n",
    "\n",
    "This insert will contain: <br>\n",
    "- PABA marker gene <br>\n",
    "- PUFA synthesis pathway <br>\n",
    "    - pfa1,2,3 and a pptase <br>\n",
    "- Hygromicin marker gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31a4f34c",
   "metadata": {},
   "source": [
    "*this is where we insert a cool image of the creation*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd194758",
   "metadata": {},
   "source": [
    "## Loading Libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff6a4f2",
   "metadata": {},
   "source": [
    "## Genetic elements"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8af7b9c4",
   "metadata": {},
   "source": [
    "### HR Domain"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd1fa971",
   "metadata": {},
   "source": [
    "The insertion site into C.Cinerea will be the spoII site based on the report: <br> *Non-conventional fungi are efficient heterologous hosts for natural product production.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38697c03",
   "metadata": {},
   "source": [
    "### PABA marker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27ad213",
   "metadata": {},
   "source": [
    "### PUFA Synthesis pathway"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aa0bead",
   "metadata": {},
   "source": [
    "### Hygromicin marker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e3b7bc",
   "metadata": {},
   "source": [
    "## Primer creation using TEEMI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4164e4d3",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0bb407a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from teemi.design.combinatorial_design import DesignAssembly\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from IPython.display import display\n",
    "import IPython.core.display\n",
    "\n",
    "import os\n",
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e4e07f",
   "metadata": {},
   "source": [
    "### Fetching everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fb531935",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Bruger\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\FastaIO.py:203: BiopythonDeprecationWarning: Previously, the FASTA parser silently ignored comments at the beginning of the FASTA file (before the first sequence).\n",
      "\n",
      "Nowadays, the FASTA file format is usually understood not to have any such comments, and most software packages do not allow them. Therefore, the use of comments at the beginning of a FASTA file is now deprecated in Biopython.\n",
      "\n",
      "In a future Biopython release, this deprecation warning will be replaced by a ValueError. To avoid this, there are three options:\n",
      "\n",
      "(1) Modify your FASTA file to remove such comments at the beginning of the file.\n",
      "\n",
      "(2) Use SeqIO.parse with the 'fasta-pearson' format instead of 'fasta'. This format is consistent with the FASTA format defined by William Pearson's FASTA aligner software. Thie format allows for comments before the first sequence; lines starting with the ';' character anywhere in the file are also regarded as comment lines and are ignored.\n",
      "\n",
      "(3) Use the 'fasta-blast' format. This format regards any lines starting with '!', '#', or ';' as comment lines. The 'fasta-blast' format may be safer than the 'fasta-pearson' format, as it explicitly indicates which lines are comments. \n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from smart_functions import read_fasta_to_dseqrecords\n",
    "\n",
    "top_promoters = r'notebooks/data/insert_sequnces/promoters.fasta'\n",
    "top_terminators = r'notebooks/data/insert_sequnces/terminators.fasta'\n",
    "\n",
    "hr_fa= r'notebooks/data/insert_sequnces/HR.fasta'\n",
    "m_paba_fa = r'notebooks/data/insert_sequnces/PABA.fasta'\n",
    "m_hygro_fa = r'notebooks/data/insert_sequnces/Hygromycin.fasta' #hygromicin including promoter and terminator.\n",
    "cds_fa = r'notebooks/data/insert_sequnces/pufa_optimized.fasta' #pfa123 and pptase\n",
    "\n",
    "promoters, promoter_names = read_fasta_to_dseqrecords(top_promoters)\n",
    "terminators, terminator_names = read_fasta_to_dseqrecords(top_terminators)\n",
    "\n",
    "cds_records, cds_names = read_fasta_to_dseqrecords(cds_fa)\n",
    "m_paba, m_paba_names = read_fasta_to_dseqrecords(m_paba_fa)\n",
    "m_hygro, m_hygro_names = read_fasta_to_dseqrecords(m_hygro_fa)\n",
    "hr_records, HR_names = read_fasta_to_dseqrecords(hr_fa)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f9ea94b",
   "metadata": {},
   "source": [
    "#### A quick count of sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "87fdf520",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Promoters: 9, CDS: 3, Terminators: 9\n"
     ]
    }
   ],
   "source": [
    "print(f\"Promoters: {len(promoters)}, CDS: {len(cds_records)}, Terminators: {len(terminators)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8854ab9",
   "metadata": {},
   "source": [
    "### Putting it all together in a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f8eeb4cd",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[12], line 13\u001b[0m\n\u001b[0;32m      9\u001b[0m     cds_block\u001b[38;5;241m.\u001b[39mextend([p, cds, t])\n\u001b[0;32m     11\u001b[0m \u001b[38;5;66;03m# Now wrap with HR and PABA components\u001b[39;00m\n\u001b[0;32m     12\u001b[0m construct \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m---> 13\u001b[0m     \u001b[43mhr_records\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m,\n\u001b[0;32m     14\u001b[0m     m_paba[\u001b[38;5;241m0\u001b[39m],\n\u001b[0;32m     15\u001b[0m     \u001b[38;5;241m*\u001b[39mcds_block,\n\u001b[0;32m     16\u001b[0m     m_paba[\u001b[38;5;241m1\u001b[39m],\n\u001b[0;32m     17\u001b[0m     hr_records[\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m     18\u001b[0m ]\n\u001b[0;32m     20\u001b[0m \u001b[38;5;66;03m# Join into final construct string\u001b[39;00m\n\u001b[0;32m     21\u001b[0m full_constructs\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(construct))\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "full_constructs = []\n",
    "\n",
    "for p in promoters:\n",
    "    for t in terminators:\n",
    "        \n",
    "        # Build the entire block of all CDSs wrapped in p and t\n",
    "        cds_block = []\n",
    "        for cds in cds_records:\n",
    "            cds_block.extend([p, cds, t])\n",
    "\n",
    "        # Now wrap with HR and PABA components\n",
    "        construct = [\n",
    "            hr_records[0],\n",
    "            m_paba[0],\n",
    "            *cds_block,\n",
    "            m_paba[1],\n",
    "            hr_records[1]\n",
    "        ]\n",
    "        \n",
    "        # Join into final construct string\n",
    "        full_constructs.append(\"-\".join(construct))\n",
    "\n",
    "print(full_constructs)\n",
    "print(len(full_constructs), \"constructs generated\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ba745a9",
   "metadata": {},
   "source": [
    "### Combining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d9a4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_TM = 55\n",
    "LIMIT = 13\n",
    "OVERLAP = 35\n",
    "\n",
    "\n",
    "design = DesignAssembly(full_constructs, list_of_pads=[], positions_of_pads=[], target_tm=TARGET_TM, limit=LIMIT, overlap=OVERLAP)\n",
    "variants_df = design.show_variants_lib_df()          \n",
    "primers_df  = design.primer_list_to_dataframe()      \n",
    "pcrs_df = design.pcr_list_to_dataframe()\n",
    "\n",
    "\n",
    "out_dir = os.path.abspath(\".\")\n",
    "variants_csv = os.path.join(out_dir, \"variants_library.csv\")\n",
    "primers_csv  = os.path.join(out_dir, \"primers_list.csv\")\n",
    "pcrs_csv= os.path.join(out_dir, \"pcr_plan.csv\")\n",
    "\n",
    "variants_df.to_csv(variants_csv, index=False)\n",
    "primers_df.to_csv(primers_csv, index=False)\n",
    "pcrs_df.to_csv(pcrs_csv, index=False)\n",
    "\n",
    "print(f\"Variants: {len(variants_df)} saved -> {variants_csv}\")\n",
    "print(f\"Primers:  {len(primers_df)} saved -> {primers_csv}\")\n",
    "print(f\"PCRs:     {len(pcrs_df)} saved -> {pcrs_csv}\")\n",
    "display(variants_df.head())\n",
    "display(primers_df.head())\n",
    "display(pcrs_df.head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
