{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bbe04a4e",
   "metadata": {},
   "source": [
    "# Creation of in vitro insertion parts for pfa chain in C.Cinerea"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d453a96d",
   "metadata": {},
   "source": [
    "Now that promoters have been collected and generated and ran through a quantative flouresence analysis we can create the full list of inserts and necesarry primers for the top 5 primers and terminators. \n",
    "\n",
    "This insert will contain: <br>\n",
    "- PABA marker gene <br>\n",
    "- PUFA synthesis pathway <br>\n",
    "    - pfa1,2,3 and a pptase <br>\n",
    "- Hygromicin marker gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31a4f34c",
   "metadata": {},
   "source": [
    "*this is where we insert a cool image of the creation*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd194758",
   "metadata": {},
   "source": [
    "## Loading Libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff6a4f2",
   "metadata": {},
   "source": [
    "## Genetic elements"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8af7b9c4",
   "metadata": {},
   "source": [
    "### HR Domain"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd1fa971",
   "metadata": {},
   "source": [
    "The insertion site into C.Cinerea will be the spoII site based on the report: <br> *Non-conventional fungi are efficient heterologous hosts for natural product production.*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38697c03",
   "metadata": {},
   "source": [
    "### PABA marker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a27ad213",
   "metadata": {},
   "source": [
    "### PUFA Synthesis pathway"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aa0bead",
   "metadata": {},
   "source": [
    "### Hygromicin marker"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e3b7bc",
   "metadata": {},
   "source": [
    "## Primer creation using TEEMI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4164e4d3",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0bb407a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from teemi.design.combinatorial_design import DesignAssembly\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from IPython.display import display\n",
    "import IPython.core.display\n",
    "\n",
    "import os\n",
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34e4e07f",
   "metadata": {},
   "source": [
    "### Fetching everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb531935",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.smart_functions import read_fasta_to_dseqrecords\n",
    "\n",
    "top_promoters = r'notebooks/data/insert_sequnces/promoters.fasta'\n",
    "top_terminators = r'notebooks/data/insert_sequnces/terminators.fasta'\n",
    "\n",
    "hr_fa= r'notebooks/data/insert_sequnces/HR.fasta'\n",
    "m_paba_fa = r'notebooks/data/insert_sequnces/PABA.fasta'\n",
    "m_hygro_fa = r'notebooks/data/insert_sequnces/Hygromycin.fasta' #hygromicin including promoter and terminator.\n",
    "cds_fa = r'notebooks/data/insert_sequnces/pufa_optimized.fasta' #pfa123 and pptase\n",
    "\n",
    "promoters, promoter_names = read_fasta_to_dseqrecords(top_promoters)\n",
    "terminators, terminator_names = read_fasta_to_dseqrecords(top_terminators)\n",
    "\n",
    "cds_records, cds_names = read_fasta_to_dseqrecords(cds_fa)\n",
    "m_paba, m_paba_names = read_fasta_to_dseqrecords(m_paba_fa)\n",
    "m_hygro, m_hygro_names = read_fasta_to_dseqrecords(m_hygro_fa)\n",
    "hr_records, HR_names = read_fasta_to_dseqrecords(hr_fa)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f9ea94b",
   "metadata": {},
   "source": [
    "#### A quick count of sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "87fdf520",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Promoters: 9, CDS: 3, Terminators: 9\n"
     ]
    }
   ],
   "source": [
    "print(f\"Promoters: {len(promoters)}, CDS: {len(cds_records)}, Terminators: {len(terminators)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8854ab9",
   "metadata": {},
   "source": [
    "### Putting it all together in a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8eeb4cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "81 constructs generated\n"
     ]
    }
   ],
   "source": [
    "full_constructs = []\n",
    "\n",
    "for p in promoters:\n",
    "    for t in terminators:\n",
    "        \n",
    "        cds_block = []\n",
    "        for cds in cds_records:\n",
    "            cds_block.extend([p, cds, t])\n",
    "\n",
    "        construct = [\n",
    "            hr_records[0],\n",
    "            m_paba[0],\n",
    "            *cds_block,\n",
    "            m_paba[1],\n",
    "            m_hygro,\n",
    "            hr_records[1]\n",
    "        ]\n",
    "        full_constructs.append(construct)\n",
    "\n",
    "print(len(full_constructs), \"constructs generated\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ba745a9",
   "metadata": {},
   "source": [
    "### Combining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d9a4d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "TARGET_TM = 55\n",
    "LIMIT = 13\n",
    "OVERLAP = 35\n",
    "\n",
    "\n",
    "design = DesignAssembly(full_constructs, list_of_pads=[], positions_of_pads=[], target_tm=TARGET_TM, limit=LIMIT, overlap=OVERLAP)\n",
    "variants_df = design.show_variants_lib_df()          \n",
    "primers_df  = design.primer_list_to_dataframe()      \n",
    "pcrs_df = design.pcr_list_to_dataframe()\n",
    "\n",
    "\n",
    "out_dir = os.path.abspath(\".\")\n",
    "variants_csv = os.path.join(out_dir, \"variants_library.csv\")\n",
    "primers_csv  = os.path.join(out_dir, \"primers_list.csv\")\n",
    "pcrs_csv= os.path.join(out_dir, \"pcr_plan.csv\")\n",
    "\n",
    "variants_df.to_csv(variants_csv, index=False)\n",
    "primers_df.to_csv(primers_csv, index=False)\n",
    "pcrs_df.to_csv(pcrs_csv, index=False)\n",
    "\n",
    "print(f\"Variants: {len(variants_df)} saved -> {variants_csv}\")\n",
    "print(f\"Primers:  {len(primers_df)} saved -> {primers_csv}\")\n",
    "print(f\"PCRs:     {len(pcrs_df)} saved -> {pcrs_csv}\")\n",
    "display(variants_df.head())\n",
    "display(primers_df.head())\n",
    "display(pcrs_df.head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
