{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4f175f5c",
   "metadata": {},
   "source": [
    "# PCR Primer design\n",
    "So far we have used Snapgene to design our primers. However in order to explore open-source alternatives this notebook will use `Biopython` in conjunction with Primer3\n",
    "## Installation\n",
    "First we need to install the python wrapper `primer3-py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "666213a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: primer3-py in c:\\users\\alexa\\anaconda3\\envs\\synbio\\lib\\site-packages (2.2.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install primer3-py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8610eda4",
   "metadata": {},
   "source": [
    "## Import\n",
    "We can now import it along with other libraries we need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3d56a86",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import primer3\n",
    "import os\n",
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06894c1",
   "metadata": {},
   "source": [
    "## Designing our first primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05c0444f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer: AGCGCACATCTCTAATCTTTGA\n",
      "Reverse primer: ACTCGCTGCCACATCACC\n",
      "Product size: 994\n"
     ]
    }
   ],
   "source": [
    "dna_seq = str(\"AGCGCACATCTCTAATCTTTGATTTACGTATAATCGTTACACAACCATCCCCCTAATTCCTCTACAGAGGCGCGTCCTGCTGCCCCTTCCCATGGCACATGCTTCCAACTTACGATTTCTCACGTCCCACTCGTCCCAGGTTTCGCCTTTTCGGATTTCGTCCGTGTTGTGTTTTGTCTTCTTCCGTTGTCTTCCTCCGTTGTCTTCGATTTGATTCCCCACACCCGACTGCCTATCATTGTTCTTGCTTCGGATTCCTGGGCCACCTCGTTGAACTGGTGCTACTCAATAATTTATCGTAATTTACCATACGTTATCGATACATACTCCCCCCCATTCCGTTGTATTCCTTCGCTGTAGCTTTTGAGACCACCACCACCACCCTAGCCTCTGTTGTATAGTTTTCGACTTCGTCCGCCCGCCGTTGTCCTTGTGAACCATTTTTGTTGCTATCGCACACATACATACCAAGTACACTGCTTCTTGCTATCCTCCCTACTAGCTTTCTACATGCATTGCTACTATACAGCCCTCTACAGCACTTGAAATGAAGTAATTTGAGCCCGGAGCATCCACGAAACCTACGCGAGGAAGAGGTTCTTTAGGCGCTGGTCTGCATTTTCGCCCATACTCAGTCCCAAATTCCAGGGAACGAGGCCAGTAATACTACTGTATTGATTGTCGTTTTACAGCCACTGTCACTGTCATTCAATGTCAACGAACGGCGTCCAGACTTGTCAAGCATCGTGCCAACGATTTGTTTCCGCGTTCACAGCATCCGCTCCAACATCTCCCTATTTTCGACGTACAAATCAAACCCTAGTTGACAGCAAACCTAACGGTAGTGTTGTTGCGAGGTGGAGGTCCCTCAAGGTGGAGTTACCTTGAGGTCCCGTATTTGTGATGGTGGCGTTCTCAAGGTCCCTTGAGGTGGCCTTGATGTGTTTTCCCATCCCTGTAACGGAATCTCCGGGACGGTGATGTGGCAGCGAGTGCGAGC\")\n",
    "\n",
    "primers = primer3.bindings.design_primers({'SEQUENCE_TEMPLATE': dna_seq,}, {'PRIMER_OPT_SIZE': 20,'PRIMER_MIN_SIZE': 18, 'PRIMER_MAX_SIZE': 25, 'PRIMER_OPT_TM': 60.0, 'PRIMER_MIN_TM': 57.0, 'PRIMER_MAX_TM': 63.0, 'PRIMER_PAIR_MAX_DIFF_TM': 3.0, 'PRIMER_PRODUCT_SIZE_RANGE': [[900, 1100]],'SEQUENCE_INCLUDED_REGION': [0, len(dna_seq)],'SEQUENCE_PRIMER_PAIR_OK_REGION_LIST': [[0, 25, len(dna_seq)-25, 25]]})\n",
    "\n",
    "# Print the first primer pair\n",
    "print(\"Forward primer:\", primers['PRIMER_LEFT_0_SEQUENCE'])\n",
    "print(\"Reverse primer:\", primers['PRIMER_RIGHT_0_SEQUENCE'])\n",
    "print(\"Product size:\", primers['PRIMER_PAIR_0_PRODUCT_SIZE'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeec4719",
   "metadata": {},
   "source": [
    "As we can see the primer is slightly shorter than the sequence due to primer3's optimization even when specifying strict positions with the last argument. In this case we choose to ignore it as it will likely not be a problem however if an important motif is left out the promoter will not work very well. To solve this one could screen the input sequence for restriction sites and choose a non-cutter restriction sequence to append to the primer and include a wider piece of the genome to ensure the entire upstream region is extracted.\n",
    "## Designing primers for genomic extraction of upstream sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "93e50fb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record: PKG1_promoter\n",
      "up primer: CGTGATTTCCTCTGCCTCGT\n",
      "down primer: ACGGAATGGTTATCGCCCTT\n",
      "product size: 935\n",
      "primr melting temp:  60.108614558536146 59.45530340378093\n",
      "\n",
      "Record: ADH1_promoter\n",
      "up primer: CAGGGTCAAGCAGAGCAGAA\n",
      "down primer: GGCACAGGGAGGTGTAAGTC\n",
      "product size: 947\n",
      "primr melting temp:  59.96411926156776 60.0362939165052\n",
      "\n",
      "Record: TDH3_promoter\n",
      "up primer: CGTTCCCACTTTGGACGTGT\n",
      "down primer: GAGAGAAAAGCGCAGTTGGC\n",
      "product size: 944\n",
      "primr melting temp:  60.812412072024244 60.11000999450579\n",
      "\n",
      "Record: ACT_promoter\n",
      "up primer: TCGAGAAGAGAGGTAGGCGG\n",
      "down primer: TGATAGAGCTGTAGGGCGGG\n",
      "product size: 950\n",
      "primr melting temp:  60.464013016904346 60.827924519505984\n",
      "\n",
      "Record: TEF1_promoter\n",
      "up primer: GGGCGAGTGTCCATTCATGA\n",
      "down primer: CATCTCGCACCAGTGGATGA\n",
      "product size: 926\n",
      "primr melting temp:  60.10770297561146 59.82355063247462\n",
      "\n",
      "Record: DED1_promoter\n",
      "up primer: CCACAGATGCAAACGCAACA\n",
      "down primer: ATCGCTGTGGATATCGGTGG\n",
      "product size: 933\n",
      "primr melting temp:  59.96949614280919 59.681559465123314\n",
      "\n",
      "Record: gpdA_promoter\n",
      "up primer: CGTTCCCACTTTGGACGTGT\n",
      "down primer: GAGAGAAAAGCGCAGTTGGC\n",
      "product size: 944\n",
      "primr melting temp:  60.812412072024244 60.11000999450579\n",
      "\n",
      "Record: pkiA_promoter\n",
      "up primer: GAGGCAATGCTGGGTTTTCC\n",
      "down primer: GTGTCCCTTTAAGTGGCGGA\n",
      "product size: 922\n",
      "primr melting temp:  59.75253990811342 59.964260505926404\n",
      "\n",
      "Record: mdhA_promoter\n",
      "up primer: CCAGTACCGCGATCCTTTGT\n",
      "down primer: GAAGGTGGTGGTTGTGGAGA\n",
      "product size: 944\n",
      "primr melting temp:  60.108329073330594 59.527529455099454\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from src.utils import design_primers\n",
    "\n",
    "pth = \"data/primer_design/promoters_for_extraction.fasta\"\n",
    "\n",
    "design_primers(pth)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3b94ea8",
   "metadata": {},
   "source": [
    "We'll also generate primers for extraction of the downstream sequences used as terminators:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "36c159fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record: PKG1_terminator\n",
      "up primer: TGATGATTATTAGTGAGAGCGTGGA\n",
      "down primer: GCCCGATACCTCCAAGGAAA\n",
      "product size: 951\n",
      "primr melting temp:  59.69819491103988 59.45530340378093\n",
      "\n",
      "Record: ADH1_terminator\n",
      "up primer: TGTGACAGATGACGGACACA\n",
      "down primer: AAGGCTTGGTTTACGACGGT\n",
      "product size: 931\n",
      "primr melting temp:  58.96329264587649 59.89273163212596\n",
      "\n",
      "Record: TDH3_terminator\n",
      "up primer: TCGTTACACAACCATCCCCC\n",
      "down primer: ACACATCAAGGCCACCTCAA\n",
      "product size: 912\n",
      "primr melting temp:  59.67439756087623 59.52192805017313\n",
      "\n",
      "Record: ACT_terminator\n",
      "up primer: ACGACCTCCTTACGACCCTT\n",
      "down primer: TCAACCAAGCCACAAGTCGT\n",
      "product size: 992\n",
      "primr melting temp:  60.25151896976945 60.107410468332944\n",
      "\n",
      "Record: TEF1_terminator\n",
      "up primer: TCACCACCTCGTTCTCGTTT\n",
      "down primer: TGACTAGGCTGCCTTTGACC\n",
      "product size: 942\n",
      "primr melting temp:  59.251549596098016 59.67505708032144\n",
      "\n",
      "Record: DED1_terminator\n",
      "up primer: GCCCCTTCTCTTTTCGACGA\n",
      "down primer: ATGTGATGCCAACATGCTGC\n",
      "product size: 938\n",
      "primr melting temp:  60.0376045227884 59.825155079357785\n",
      "\n",
      "Record: gpdA_terminator\n",
      "up primer: CGTTCCCACTTTGGACGTGT\n",
      "down primer: GAGAGAAAAGCGCAGTTGGC\n",
      "product size: 944\n",
      "primr melting temp:  60.812412072024244 60.11000999450579\n",
      "\n",
      "Record: pkiA_terminator\n",
      "up primer: GCTTTCGCCATTCTACTCGC\n",
      "down primer: CCCTTGCCTGTCTATCGACC\n",
      "product size: 920\n",
      "primr melting temp:  59.69801256000028 59.89496080906696\n",
      "\n",
      "Record: mdhA_terminator\n",
      "up primer: GGCGGGTGGTTAGATGGTAG\n",
      "down primer: CCGATTTACCTCTCCCAGCG\n",
      "product size: 916\n",
      "primr melting temp:  59.89385051200492 60.249488975504505\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pth = \"data/primer_design/terminators_for_extraction.fasta\"\n",
    "\n",
    "design_primers(pth)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a6b70d5",
   "metadata": {},
   "source": [
    "### A very specific function for generating a table in our report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4f33682",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\textit{PKG1_promoter} & Up & 60 & 5'-CGTGATTTCCTCTGCCTCGT-3' \\\\\n",
      "& Down & 59 & 5'-ACGGAATGGTTATCGCCCTT-3' \\\\\n",
      "\\textit{ADH1_promoter} & Up & 60 & 5'-CAGGGTCAAGCAGAGCAGAA-3' \\\\\n",
      "& Down & 60 & 5'-GGCACAGGGAGGTGTAAGTC-3' \\\\\n",
      "\\textit{TDH3_promoter} & Up & 61 & 5'-CGTTCCCACTTTGGACGTGT-3' \\\\\n",
      "& Down & 60 & 5'-GAGAGAAAAGCGCAGTTGGC-3' \\\\\n",
      "\\textit{ACT_promoter} & Up & 60 & 5'-TCGAGAAGAGAGGTAGGCGG-3' \\\\\n",
      "& Down & 61 & 5'-TGATAGAGCTGTAGGGCGGG-3' \\\\\n",
      "\\textit{TEF1_promoter} & Up & 60 & 5'-GGGCGAGTGTCCATTCATGA-3' \\\\\n",
      "& Down & 60 & 5'-CATCTCGCACCAGTGGATGA-3' \\\\\n",
      "\\textit{DED1_promoter} & Up & 60 & 5'-CCACAGATGCAAACGCAACA-3' \\\\\n",
      "& Down & 60 & 5'-ATCGCTGTGGATATCGGTGG-3' \\\\\n",
      "\\textit{gpdA_promoter} & Up & 61 & 5'-CGTTCCCACTTTGGACGTGT-3' \\\\\n",
      "& Down & 60 & 5'-GAGAGAAAAGCGCAGTTGGC-3' \\\\\n",
      "\\textit{pkiA_promoter} & Up & 60 & 5'-GAGGCAATGCTGGGTTTTCC-3' \\\\\n",
      "& Down & 60 & 5'-GTGTCCCTTTAAGTGGCGGA-3' \\\\\n",
      "\\textit{mdhA_promoter} & Up & 60 & 5'-CCAGTACCGCGATCCTTTGT-3' \\\\\n",
      "& Down & 60 & 5'-GAAGGTGGTGGTTGTGGAGA-3' \\\\\n"
     ]
    }
   ],
   "source": [
    "from src.utils import design_primers_latex_table\n",
    "#promoters\n",
    "design_primers_latex_table(\"data/primer_design/promoters_for_extraction.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a823e65d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\textit{PKG1_terminator} & Up & 60 & 5'-TGATGATTATTAGTGAGAGCGTGGA-3' \\\\\n",
      "& Down & 59 & 5'-GCCCGATACCTCCAAGGAAA-3' \\\\\n",
      "\\textit{ADH1_terminator} & Up & 59 & 5'-TGTGACAGATGACGGACACA-3' \\\\\n",
      "& Down & 60 & 5'-AAGGCTTGGTTTACGACGGT-3' \\\\\n",
      "\\textit{TDH3_terminator} & Up & 60 & 5'-TCGTTACACAACCATCCCCC-3' \\\\\n",
      "& Down & 60 & 5'-ACACATCAAGGCCACCTCAA-3' \\\\\n",
      "\\textit{ACT_terminator} & Up & 60 & 5'-ACGACCTCCTTACGACCCTT-3' \\\\\n",
      "& Down & 60 & 5'-TCAACCAAGCCACAAGTCGT-3' \\\\\n",
      "\\textit{TEF1_terminator} & Up & 59 & 5'-TCACCACCTCGTTCTCGTTT-3' \\\\\n",
      "& Down & 60 & 5'-TGACTAGGCTGCCTTTGACC-3' \\\\\n",
      "\\textit{DED1_terminator} & Up & 60 & 5'-GCCCCTTCTCTTTTCGACGA-3' \\\\\n",
      "& Down & 60 & 5'-ATGTGATGCCAACATGCTGC-3' \\\\\n",
      "\\textit{gpdA_terminator} & Up & 61 & 5'-CGTTCCCACTTTGGACGTGT-3' \\\\\n",
      "& Down & 60 & 5'-GAGAGAAAAGCGCAGTTGGC-3' \\\\\n",
      "\\textit{pkiA_terminator} & Up & 60 & 5'-GCTTTCGCCATTCTACTCGC-3' \\\\\n",
      "& Down & 60 & 5'-CCCTTGCCTGTCTATCGACC-3' \\\\\n",
      "\\textit{mdhA_terminator} & Up & 60 & 5'-GGCGGGTGGTTAGATGGTAG-3' \\\\\n",
      "& Down & 60 & 5'-CCGATTTACCTCTCCCAGCG-3' \\\\\n"
     ]
    }
   ],
   "source": [
    "#terminators\n",
    "pth = \"data/primer_design/terminators_for_extraction.fasta\"\n",
    "design_primers_latex_table(pth)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16196015",
   "metadata": {},
   "source": [
    "## Generating primers for PCR validation of construct\n",
    "To validate the inserts with PCR, we need primers annealing to known sequences on each promoter and terminator. This way, we can easily verify whether the assembly was successfull by combining the products with restriction enzymes and running on gel. We can then assess whether the observed lengths match our expectations. \n",
    "### Full construct validation\n",
    "Each promoter and terminator needs its own primer. We design the primers so they anneal at different lengths from the end of promoter. This allows us to verify which promoter or terminator was inserted by checking the length of the resulting bands. Furthermore we need both forward and reverse primers at each CDS and a forward primer on the upstream PABA and a reverse primer on the downstream PABA. With this assembly we are able to verify each step in the final assembly. Please see the figure below:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d576d8a",
   "metadata": {},
   "source": [
    "![alt text](primers.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28fc667",
   "metadata": {},
   "source": [
    "Note that we will also need reverse primers on promoters and forward primers on terminators to verify assembly of the PABA-ends. Also note that on the figure the CDS primers are shown near the middle. In reality they are placed near the end to keep the resulting products small to make the difference distinguisable.\n",
    "### Primers for promoters\n",
    "Each promoter needs a forward promoter annealing at varying positions. The reverse promoter does not need to vary in position as the promoter can be identified via the product between promoter and CDS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "60c7a9c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N primers: 19\n"
     ]
    }
   ],
   "source": [
    "from src.utils import create_end_primers\n",
    "\n",
    "# generate the primers we need for promoters  \n",
    "primers = create_end_primers(\"data\\promoter_terminator_library\\promoter_library.fasta\", out_fasta=\"data/primer_design/promoters_end_primers.fasta\", primer_len=30, first_offset=20, step=30, reverse_complement=False)\n",
    "total_records = len(primers)\n",
    "print(f\"N primers: {total_records}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a09c8d46",
   "metadata": {},
   "source": [
    "We also need reverse primers for the promoters. For these we can simply take the reverse compliment to the start of each promoter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "55c374ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N primers: 19\n"
     ]
    }
   ],
   "source": [
    "read_file = \"data\\promoter_terminator_library\\promoter_library.fasta\"\n",
    "to_file = \"data/primer_design/promoters_reverse_primers.fasta\"\n",
    "\n",
    "primers = create_end_primers(read_file, out_fasta=to_file, primer_len=30, first_offset=950, step=30, reverse_complement=True)\n",
    "print(f\"N primers: {len(primers)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b89ac51a",
   "metadata": {},
   "source": [
    "### Terminator primers\n",
    "The above procedure is repeated for the terminators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0e54b98f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N reverse primers: 9\n",
      "N forward primers: 9\n"
     ]
    }
   ],
   "source": [
    "read_file = \"data/promoter_terminator_library/terminator_library.fasta\"\n",
    "to_file = \"data/primer_design/terminators_reverse_primers.fasta\"\n",
    "\n",
    "primers = create_end_primers(read_file, out_fasta=to_file, primer_len=30, first_offset=930, step=-30, reverse_complement=True)\n",
    "print(f\"N reverse primers: {len(primers)}\")\n",
    "\n",
    "#forward primers for terminators\n",
    "read_file = \"data/promoter_terminator_library/terminator_library.fasta\"\n",
    "to_file = \"data/primer_design/terminators_forward_primers.fasta\"\n",
    "\n",
    "primers = create_end_primers(read_file, out_fasta=to_file, primer_len=30, first_offset=20, step=30, reverse_complement=False)\n",
    "print(f\"N forward primers: {len(primers)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c7aa8d5",
   "metadata": {},
   "source": [
    "### PABA primers\n",
    "We need the PABA primers:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "343529fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PABA upstream forward primer: TGATAGGTCGTAGCTGGCACACAGAATGAG\n",
      "PABA downstream reverse primer: GAAGCGACATTTGGGATCAGGAGAGTGGCC\n"
     ]
    }
   ],
   "source": [
    "read_file = \"data/insert_sequences/PABA.fasta\"\n",
    "\n",
    "#reading to seq object\n",
    "records = list(SeqIO.parse(read_file, \"fasta\"))\n",
    "paba_rec_1 = records[0]\n",
    "paba_up = paba_rec_1.seq\n",
    "paba_rec = records[1]\n",
    "paba_down = paba_rec.seq\n",
    "\n",
    "# we need a forward primer for the upstream PABA:\n",
    "paba_up_forward_primer = paba_up[-50:-20]  # last 30 bases of the upstream sequence\n",
    "print(\"PABA upstream forward primer:\", paba_up_forward_primer)\n",
    "\n",
    "# and a reverse primer for the downstream PABA:\n",
    "paba_down_reverse_primer = paba_down[20:50].reverse_complement()\n",
    "print(\"PABA downstream reverse primer:\", paba_down_reverse_primer)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9374e34a",
   "metadata": {},
   "source": [
    "## dTomato validation\n",
    "Finally we need dTomato (and negative control CDS) forward and reverse primers for validation of the dTomato variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "157f045b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dTomato forward primer: GGTCATCAAAGAGTTCATGCGCTTCAAGGT\n",
      "dTomato reverse primer: CGTACAGGAACAGGTGGTGGCGGCCCTCGG\n",
      "Control forward primer: GCACAGCTCCGACGAGTTCGGAATAGAGAC\n",
      "Control reverse primer: GGGTCTCCACGTTTCCGGTCAACGGATGCA\n"
     ]
    }
   ],
   "source": [
    "read_file = \"data/insert_sequences/dTomato_non_optimized.fasta\"\n",
    "\n",
    "#reading to seq objects\n",
    "records = list(SeqIO.parse(read_file, \"fasta\"))\n",
    "dTomato_rec = records[0]\n",
    "dTomato = dTomato_rec.seq\n",
    "non_fluo_rec = records[1]\n",
    "non_fluo = non_fluo_rec.seq\n",
    "\n",
    "# forward and reverse for dTomato\n",
    "dTomato_forward_primer = dTomato[20:50]\n",
    "dTomato_reverse_primer = dTomato[-50:-20].reverse_complement()\n",
    "print(\"dTomato forward primer:\", dTomato_forward_primer)\n",
    "print(\"dTomato reverse primer:\", dTomato_reverse_primer)\n",
    "# forward and reverse for non-fluorescent control\n",
    "non_fluo_forward_primer = non_fluo[20:50]\n",
    "non_fluo_reverse_primer = non_fluo[-50:-20].reverse_complement()\n",
    "\n",
    "print(\"Control forward primer:\", non_fluo_forward_primer)\n",
    "print(\"Control reverse primer:\", non_fluo_reverse_primer)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "synbio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
