{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4f175f5c",
   "metadata": {},
   "source": [
    "# PCR Primer design\n",
    "So far we have used Snapgene to design our primers. However in order to explore open-source alternatives this notebook will use `Biopython` in conjunction with Primer3\n",
    "## Installation\n",
    "First we need to install the python wrapper `primer3-py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "666213a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting primer3-py\n",
      "  Downloading primer3_py-2.2.0-cp311-cp311-win_amd64.whl.metadata (4.1 kB)\n",
      "Downloading primer3_py-2.2.0-cp311-cp311-win_amd64.whl (1.6 MB)\n",
      "   ---------------------------------------- 0.0/1.6 MB ? eta -:--:--\n",
      "   -------------------------- ------------- 1.0/1.6 MB ? eta -:--:--\n",
      "   -------------------------- ------------- 1.0/1.6 MB ? eta -:--:--\n",
      "   -------------------------- ------------- 1.0/1.6 MB ? eta -:--:--\n",
      "   -------------------------- ------------- 1.0/1.6 MB ? eta -:--:--\n",
      "   ---------------------------------------- 1.6/1.6 MB 1.3 MB/s  0:00:01\n",
      "Installing collected packages: primer3-py\n",
      "Successfully installed primer3-py-2.2.0\n"
     ]
    }
   ],
   "source": [
    "!pip install primer3-py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8610eda4",
   "metadata": {},
   "source": [
    "## Import\n",
    "We can now import it along with other libraries we need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a3d56a86",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import primer3\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06894c1",
   "metadata": {},
   "source": [
    "## Designing our first primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05c0444f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer: AGCGCACATCTCTAATCTTTGA\n",
      "Reverse primer: ACTCGCTGCCACATCACC\n",
      "Product size: 994\n"
     ]
    }
   ],
   "source": [
    "dna_seq = str(\"AGCGCACATCTCTAATCTTTGATTTACGTATAATCGTTACACAACCATCCCCCTAATTCCTCTACAGAGGCGCGTCCTGCTGCCCCTTCCCATGGCACATGCTTCCAACTTACGATTTCTCACGTCCCACTCGTCCCAGGTTTCGCCTTTTCGGATTTCGTCCGTGTTGTGTTTTGTCTTCTTCCGTTGTCTTCCTCCGTTGTCTTCGATTTGATTCCCCACACCCGACTGCCTATCATTGTTCTTGCTTCGGATTCCTGGGCCACCTCGTTGAACTGGTGCTACTCAATAATTTATCGTAATTTACCATACGTTATCGATACATACTCCCCCCCATTCCGTTGTATTCCTTCGCTGTAGCTTTTGAGACCACCACCACCACCCTAGCCTCTGTTGTATAGTTTTCGACTTCGTCCGCCCGCCGTTGTCCTTGTGAACCATTTTTGTTGCTATCGCACACATACATACCAAGTACACTGCTTCTTGCTATCCTCCCTACTAGCTTTCTACATGCATTGCTACTATACAGCCCTCTACAGCACTTGAAATGAAGTAATTTGAGCCCGGAGCATCCACGAAACCTACGCGAGGAAGAGGTTCTTTAGGCGCTGGTCTGCATTTTCGCCCATACTCAGTCCCAAATTCCAGGGAACGAGGCCAGTAATACTACTGTATTGATTGTCGTTTTACAGCCACTGTCACTGTCATTCAATGTCAACGAACGGCGTCCAGACTTGTCAAGCATCGTGCCAACGATTTGTTTCCGCGTTCACAGCATCCGCTCCAACATCTCCCTATTTTCGACGTACAAATCAAACCCTAGTTGACAGCAAACCTAACGGTAGTGTTGTTGCGAGGTGGAGGTCCCTCAAGGTGGAGTTACCTTGAGGTCCCGTATTTGTGATGGTGGCGTTCTCAAGGTCCCTTGAGGTGGCCTTGATGTGTTTTCCCATCCCTGTAACGGAATCTCCGGGACGGTGATGTGGCAGCGAGTGCGAGC\")\n",
    "\n",
    "primers = primer3.bindings.design_primers({'SEQUENCE_TEMPLATE': dna_seq,}, {'PRIMER_OPT_SIZE': 20,'PRIMER_MIN_SIZE': 18, 'PRIMER_MAX_SIZE': 25, 'PRIMER_OPT_TM': 60.0, 'PRIMER_MIN_TM': 57.0, 'PRIMER_MAX_TM': 63.0, 'PRIMER_PAIR_MAX_DIFF_TM': 3.0, 'PRIMER_PRODUCT_SIZE_RANGE': [[900, 1100]],'SEQUENCE_INCLUDED_REGION': [0, len(dna_seq)],'SEQUENCE_PRIMER_PAIR_OK_REGION_LIST': [[0, 25, len(dna_seq)-25, 25]]})\n",
    "\n",
    "# Print the first primer pair\n",
    "print(\"Forward primer:\", primers['PRIMER_LEFT_0_SEQUENCE'])\n",
    "print(\"Reverse primer:\", primers['PRIMER_RIGHT_0_SEQUENCE'])\n",
    "print(\"Product size:\", primers['PRIMER_PAIR_0_PRODUCT_SIZE'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeec4719",
   "metadata": {},
   "source": [
    "As we can see the primer is slightly shorter than the sequence due to primer3's optimization even when specifying strict positions with the last argument. In this case we choose to ignore it as it will likely not be a problem however if an important motif is left out the promoter will not work very well. To solve this one could screen the input sequence for restriction sites and choose a non-cutter restriction sequence to append to the primer and include a wider piece of the genome to ensure the entire upstream region is extracted."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "synbio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
