{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4f175f5c",
   "metadata": {},
   "source": [
    "# PCR Primer design\n",
    "So far we have used Snapgene to design our primers. However in order to explore open-source alternatives this notebook will use `Biopython` in conjunction with Primer3\n",
    "## Installation\n",
    "First we need to install the python wrapper `primer3-py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "666213a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: primer3-py in c:\\users\\alexa\\anaconda3\\envs\\synbio\\lib\\site-packages (2.2.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install primer3-py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8610eda4",
   "metadata": {},
   "source": [
    "## Import\n",
    "We can now import it along with other libraries we need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3d56a86",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import primer3\n",
    "import os\n",
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b06894c1",
   "metadata": {},
   "source": [
    "## Designing our first primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "05c0444f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer: AGCGCACATCTCTAATCTTTGA\n",
      "Reverse primer: ACTCGCTGCCACATCACC\n",
      "Product size: 994\n"
     ]
    }
   ],
   "source": [
    "dna_seq = str(\"AGCGCACATCTCTAATCTTTGATTTACGTATAATCGTTACACAACCATCCCCCTAATTCCTCTACAGAGGCGCGTCCTGCTGCCCCTTCCCATGGCACATGCTTCCAACTTACGATTTCTCACGTCCCACTCGTCCCAGGTTTCGCCTTTTCGGATTTCGTCCGTGTTGTGTTTTGTCTTCTTCCGTTGTCTTCCTCCGTTGTCTTCGATTTGATTCCCCACACCCGACTGCCTATCATTGTTCTTGCTTCGGATTCCTGGGCCACCTCGTTGAACTGGTGCTACTCAATAATTTATCGTAATTTACCATACGTTATCGATACATACTCCCCCCCATTCCGTTGTATTCCTTCGCTGTAGCTTTTGAGACCACCACCACCACCCTAGCCTCTGTTGTATAGTTTTCGACTTCGTCCGCCCGCCGTTGTCCTTGTGAACCATTTTTGTTGCTATCGCACACATACATACCAAGTACACTGCTTCTTGCTATCCTCCCTACTAGCTTTCTACATGCATTGCTACTATACAGCCCTCTACAGCACTTGAAATGAAGTAATTTGAGCCCGGAGCATCCACGAAACCTACGCGAGGAAGAGGTTCTTTAGGCGCTGGTCTGCATTTTCGCCCATACTCAGTCCCAAATTCCAGGGAACGAGGCCAGTAATACTACTGTATTGATTGTCGTTTTACAGCCACTGTCACTGTCATTCAATGTCAACGAACGGCGTCCAGACTTGTCAAGCATCGTGCCAACGATTTGTTTCCGCGTTCACAGCATCCGCTCCAACATCTCCCTATTTTCGACGTACAAATCAAACCCTAGTTGACAGCAAACCTAACGGTAGTGTTGTTGCGAGGTGGAGGTCCCTCAAGGTGGAGTTACCTTGAGGTCCCGTATTTGTGATGGTGGCGTTCTCAAGGTCCCTTGAGGTGGCCTTGATGTGTTTTCCCATCCCTGTAACGGAATCTCCGGGACGGTGATGTGGCAGCGAGTGCGAGC\")\n",
    "\n",
    "primers = primer3.bindings.design_primers({'SEQUENCE_TEMPLATE': dna_seq,}, {'PRIMER_OPT_SIZE': 20,'PRIMER_MIN_SIZE': 18, 'PRIMER_MAX_SIZE': 25, 'PRIMER_OPT_TM': 60.0, 'PRIMER_MIN_TM': 57.0, 'PRIMER_MAX_TM': 63.0, 'PRIMER_PAIR_MAX_DIFF_TM': 3.0, 'PRIMER_PRODUCT_SIZE_RANGE': [[900, 1100]],'SEQUENCE_INCLUDED_REGION': [0, len(dna_seq)],'SEQUENCE_PRIMER_PAIR_OK_REGION_LIST': [[0, 25, len(dna_seq)-25, 25]]})\n",
    "\n",
    "# Print the first primer pair\n",
    "print(\"Forward primer:\", primers['PRIMER_LEFT_0_SEQUENCE'])\n",
    "print(\"Reverse primer:\", primers['PRIMER_RIGHT_0_SEQUENCE'])\n",
    "print(\"Product size:\", primers['PRIMER_PAIR_0_PRODUCT_SIZE'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeec4719",
   "metadata": {},
   "source": [
    "As we can see the primer is slightly shorter than the sequence due to primer3's optimization even when specifying strict positions with the last argument. In this case we choose to ignore it as it will likely not be a problem however if an important motif is left out the promoter will not work very well. To solve this one could screen the input sequence for restriction sites and choose a non-cutter restriction sequence to append to the primer and include a wider piece of the genome to ensure the entire upstream region is extracted.\n",
    "## Designing primers for genomic extraction of upstream sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "93e50fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pth = \"data\\primer_design\\promoters_for_extraction.fasta\"\n",
    "\n",
    "def design_primers(pth):\n",
    "    for record in SeqIO.parse(pth, \"fasta\"):\n",
    "        dna_seq = str(record.seq)\n",
    "        primers = primer3.bindings.design_primers({'SEQUENCE_TEMPLATE': dna_seq,}, {'PRIMER_OPT_SIZE': 20,'PRIMER_MIN_SIZE': 18, 'PRIMER_MAX_SIZE': 25, 'PRIMER_OPT_TM': 60.0, 'PRIMER_MIN_TM': 57.0, 'PRIMER_MAX_TM': 63.0, 'PRIMER_PAIR_MAX_DIFF_TM': 3.0, 'PRIMER_PRODUCT_SIZE_RANGE': [[900, 1100]],'SEQUENCE_INCLUDED_REGION': [0, len(dna_seq)],'SEQUENCE_PRIMER_PAIR_OK_REGION_LIST': [[0, 50, len(dna_seq)-50, 50]]})\n",
    "        print(f\"Record: {record.id}\")\n",
    "        print(\"up primer:\", primers['PRIMER_LEFT_0_SEQUENCE'])\n",
    "        print(\"down primer:\", primers['PRIMER_RIGHT_0_SEQUENCE'])\n",
    "        print(\"product size:\", primers['PRIMER_PAIR_0_PRODUCT_SIZE'])\n",
    "        print(\"primr melting temp: \", primers['PRIMER_LEFT_0_TM'], primers['PRIMER_RIGHT_0_TM'])\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c0804a93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record: PKG1_promoter\n",
      "up primer: ATCTGGTCGTGATTCGCCAC\n",
      "down primer: TCGTGAAGGAACGGAATGGT\n",
      "product size: 971\n",
      "primr melting temp:  60.46028408477713 59.317148410654\n",
      "\n",
      "Record: ADH1_promoter\n",
      "up primer: CAGGGTCAAGCAGAGCAGAA\n",
      "down primer: GGTTGATGCTGAGTGTGATCG\n",
      "product size: 989\n",
      "primr melting temp:  59.96411926156776 59.33737138101719\n",
      "\n",
      "Record: TDH3_promoter\n",
      "up primer: CGTTCCCACTTTGGACGTGT\n",
      "down primer: TGTGGTGGATTGGGTGATGA\n",
      "product size: 975\n",
      "primr melting temp:  60.812412072024244 58.928565557755405\n",
      "\n",
      "Record: ACT_promoter\n",
      "up primer: TCGAGAAGAGAGGTAGGCGG\n",
      "down primer: AAGGAAGAAAACTGGTGGTTGA\n",
      "product size: 969\n",
      "primr melting temp:  60.464013016904346 58.02398523353452\n",
      "\n",
      "Record: TEF1_promoter\n",
      "up primer: TTATGTTCTCCTCGTGGGCG\n",
      "down primer: ATCGTCAGCATCCATCTCGC\n",
      "product size: 953\n",
      "primr melting temp:  59.825155079357785 60.31938261960141\n",
      "\n",
      "Record: DED1_promoter\n",
      "up primer: CCACAGATGCAAACGCAACA\n",
      "down primer: TCGTTCTTGAATCGCTGTGGA\n",
      "product size: 943\n",
      "primr melting temp:  59.96949614280919 59.99960552843305\n",
      "\n",
      "Record: gpdA_promoter\n",
      "up primer: CGTTCCCACTTTGGACGTGT\n",
      "down primer: TGTGGTGGATTGGGTGATGA\n",
      "product size: 975\n",
      "primr melting temp:  60.812412072024244 58.928565557755405\n",
      "\n",
      "Record: pkiA_promoter\n",
      "up primer: GAGGCAATGCTGGGTTTTCC\n",
      "down primer: GAAGTTGGGTGTGCAGAGGA\n",
      "product size: 964\n",
      "primr melting temp:  59.75253990811342 59.89141529116807\n",
      "\n",
      "Record: mdhA_promoter\n",
      "up primer: CCAGTACCGCGATCCTTTGT\n",
      "down primer: TGAAGAGGAGAAGGTGGTGG\n",
      "product size: 953\n",
      "primr melting temp:  60.108329073330594 58.647916876715044\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from src.utils import design_primers\n",
    "\n",
    "design_primers(pth)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "synbio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
